Calendar Plugin

fatihak opened the pull request:
Adds a calendar plugin that supports displaying events from Google Calendar, Microsoft Outlook, and Apple Calendar via iCal URLs. Includes multiple layout options: day, week, month, and list.

You'll need to checkout the branch and run the [update script](https://github.com/fatihak/InkyPi?tab=readme-ov-file#update) to test. Let me know if there are any issues!


![Screenshot 2025-06-18 at 10 22 27 PM](https://github.com/user-attachments/assets/0449f838-d3f0-4803-b9ef-93e619ae85b8)
![current_image (1)](https://github.com/user-attachments/assets/6597863e-0cce-4743-a9ee-fa4ad71364f1)

phattmatt commented:
Amazing 😁

Initial testing looks good:

- Upgrade from an existing installation worked fine.
- Adding a Calendar from a private Google Calendar, using the "Integrate calendar \ Secret address in iCal format" feature, as a one time "Update Now" worked fine.
- Adding two Calendars (from Google, using the same option as above) as a playlist instance worked fine.

Will continue to play.

Great work 👍

phattmatt commented:
> Wanted to test this out, but have been hitting an error when trying to load the calendar plugin page in the UI:
> 
> `"error": "An error occurred: Plugin 'calendar' is not registered."`

What process have you followed so far?

It should look something like:

1. Log into the Raspberry Pi over SSH
2. Stop the InkyPi service: `sudo systemctl stop inkypi`
3. Changed to the InkyPi root folder: `pi@inkypi-dev:~ $ cd ./InkyPi/`
4. Run `git fetch` to get latest remote repository content
5. Run `git switch -c origin/calendar_plugin` to change to the Calendar Branch
6. Run `sudo install/update.sh` to install any new dependencies 
7. Start the InkyPi service: `sudo systemctl start inkypi`

jevidon commented:
First of all, a huge thank you @fatihak for this project and specifically for the calendar plugin. This is exactly the use case that had me getting the 13" inky display in the first place. I would say that as is this is a great implementation, however there are a few bits of UX feedback I would like to offer:

- i see an option to set the text color of the overall frame and labels, but it would be great to have a way to set the font color for each calendar alongside the calendar color selection (i believe it's white only, which makes lighter color backgrounds for the events hard to read)
- while it is nice to have a hex color selector to customize the color for each calendar, many colors don't render very cleanly (e.g. lots of dots like shading almost) which make it harder to read the text. i'm going to do some trial and error to see which colors work best for the e-ink screen, but maybe we can add in some suggested presets for color choices? 
- the "week" layout is likely the choice that I would personally use, however unlike on a computer I can't quickly jump ahead to look at the next week ahead.  it would be great to offer a "X days" ahead where today is always the first column on screen (similar to the "7 days" layout choice in Google Calendar)
- i may end up setting the update interval to be every hour, i'm wondering if an optional visual reference to show what box is the current hour might be a nice touch, similar to the bold box around the day column header.

Looking forward to seeing what else gets built for this project as it continues to evolve!

fatihak commented:
Thanks for testing @phattmatt @jevidon !

I pushed a few improvements based on some of your feedback:
- Added a Display "Now Indicator" option that displays an red arrow and line at the top of the current hour
- Added a Display "Previous Days" option that lets you hide previous days on a week layout so the current day is the first day when enabled

Regarding some of the other points:
- I'm hesitant to add additional color options/customizations to keep things simple, I feel like existing options you should be able to set the calendar colors with some trial and error to improve contrast
- I don't think there's a good way to handle shorter events. You can try increasing the overall area of the calendar by hiding event times, setting a shorter time interval, or hiding the title. 

You'll need to run `git pull` and `sudo systemctl start inkypi` to see the latest changes

jevidon commented:
Agreed with most of the hesitation you expressed @fatihak although I think adding the ability to define the text (even if just a toggle between black and white) would be a very useful addition.

I'll get this latest iteration loaded and tested later today.

jevidon commented:
Just tested the Weekly layout with "Now Indicator" enabled and "Previous Days" turned off:

- Appears that the ical pull is not grabbing any events from next week, only populates events listed up through end of day Sunday
- The red indicator could be a touch thicker IMO, but it's a nice touch as is

phattmatt commented:
> Just updated, but not seeing next week's events loading in.

Weird, I am (Google Calendar). 

Did you restart (or stop and start) the inkypi service after running the git pull? E.g:

sudo systemctl restart inkypi

jevidon commented:
Did another restart and they are now showing! Odd since I was seeing some of the other updates that came through with those latest commits.

Thank you for the assistance @phattmatt - I'm not a developer, so some of this stuff doesn't come as naturally for me.

fatihak commented:
Most code changes require a restart unless its just an a change to an html/css file.
Just pushed another change that auto assigns black/white text color based on the calendar color so it should be a bit better.

phattmatt commented:
Played around some more:

Added a Hotmail/Outlook.com calendar and this also seems to be working flawlessly.

I have 2 x Google and 1x Hotmail configured, it that are all displayed fine.

Not encountered any issues so far.

jevidon commented:
- ran `git pull origin calendar_plugin` followed by `sudo systemctl restart inkypi`
- rebooted within the web UI
- updated the font size to "large"
- manually triggered a refresh of the playlist

Indicated in the UI that it successfully initiated the refresh but the screen has not updated. Also tried switching to a basic image display, but that also does not trigger a screen update.

fatihak commented:
Can you tail the logs with this command, trigger a refresh, and check if there are any errors?
journalctl -u inkypi -f

jevidon commented:
> Can you tail the logs with this command, trigger a refresh, and check if there are any errors? journalctl -u inkypi -f

`Jun 23 19:14:56 jevipi inkypi[767]: 19:14:56 - DEBUG - config - Getting key timezone from config.
Jun 23 19:14:56 jevipi inkypi[767]: 19:14:56 - INFO - refresh_task - Running interval refresh check. | current_time: 2025-06-23 19:14:56
Jun 23 19:14:56 jevipi inkypi[767]: 19:14:56 - DEBUG - config - Getting key plugin_cycle_interval_seconds from config.
Jun 23 19:14:56 jevipi inkypi[767]: 19:14:56 - INFO - refresh_task - Not time to update display. | latest_update: 2025-06-23 18:48:40 | plugin_cycle_interval: 3600
Jun 23 19:14:56 jevipi inkypi[767]: 19:14:56 - DEBUG - config - Getting key scheduler_sleep_time from config.
Jun 23 19:15:56 jevipi inkypi[767]: 19:15:56 - DEBUG - config - Getting key timezone from config.
Jun 23 19:15:56 jevipi inkypi[767]: 19:15:56 - INFO - refresh_task - Running interval refresh check. | current_time: 2025-06-23 19:15:56
Jun 23 19:15:56 jevipi inkypi[767]: 19:15:56 - DEBUG - config - Getting key plugin_cycle_interval_seconds from config.
Jun 23 19:15:56 jevipi inkypi[767]: 19:15:56 - INFO - refresh_task - Not time to update display. | latest_update: 2025-06-23 18:48:40 | plugin_cycle_interval: 3600
Jun 23 19:15:56 jevipi inkypi[767]: 19:15:56 - DEBUG - config - Getting key scheduler_sleep_time from config.
Jun 23 19:16:51 jevipi inkypi[767]: 19:16:51 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:51] "GET /playlist HTTP/1.1" 200 -
Jun 23 19:16:52 jevipi inkypi[767]: 19:16:52 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:52] "GET /static/styles/main.css HTTP/1.1" 304 -
Jun 23 19:16:52 jevipi inkypi[767]: 19:16:52 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:52] "GET /static/scripts/response_modal.js HTTP/1.1" 304 -
Jun 23 19:16:52 jevipi inkypi[767]: 19:16:52 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:52] "GET /static/icons/playlist.png HTTP/1.1" 304 -
Jun 23 19:16:52 jevipi inkypi[767]: 19:16:52 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:52] "GET /static/icons/settings.png HTTP/1.1" 304 -
Jun 23 19:16:52 jevipi inkypi[767]: 19:16:52 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:52] "GET /images/calendar/icon.png HTTP/1.1" 304 -
Jun 23 19:16:52 jevipi inkypi[767]: 19:16:52 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:52] "GET /static/icons/edit.png HTTP/1.1" 304 -
Jun 23 19:16:52 jevipi inkypi[767]: 19:16:52 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:52] "GET /static/icons/remove.png HTTP/1.1" 304 -
Jun 23 19:16:52 jevipi inkypi[767]: 19:16:52 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:16:52] "GET /static/icons/display.png HTTP/1.1" 304 -
Jun 23 19:16:54 jevipi inkypi[767]: 19:16:54 - DEBUG - config - Getting key timezone from config.
Jun 23 19:16:54 jevipi inkypi[767]: 19:16:54 - INFO - refresh_task - Manual update requested
Jun 23 19:16:54 jevipi inkypi[767]: 19:16:54 - INFO - refresh_task - Refreshing plugin instance. | plugin_instance: 'Now'
Jun 23 19:16:54 jevipi inkypi[767]: 19:16:54 - DEBUG - config - Getting key resolution from config.
Jun 23 19:16:54 jevipi inkypi[767]: 19:16:54 - DEBUG - config - Getting key orientation from config.
Jun 23 19:16:54 jevipi inkypi[767]: 19:16:54 - DEBUG - config - Getting key timezone from config.
Jun 23 19:16:54 jevipi inkypi[767]: 19:16:54 - DEBUG - config - Getting key time_format from config.
Jun 23 19:16:54 jevipi inkypi[767]: 19:16:54 - DEBUG - urllib3.connectionpool - Starting new HTTPS connection (1): calendar.google.com:443
Jun 23 19:16:55 jevipi inkypi[767]: 19:16:55 - DEBUG - urllib3.connectionpool - https://calendar.google.com:443 "GET [redacted] HTTP/1.1" 200 None
Jun 23 19:17:53 jevipi inkypi[767]: 19:17:53 - DEBUG - PIL.PngImagePlugin - STREAM b'IHDR' 16 13
Jun 23 19:17:53 jevipi inkypi[767]: 19:17:53 - DEBUG - PIL.PngImagePlugin - STREAM b'sRGB' 41 1
Jun 23 19:17:53 jevipi inkypi[767]: 19:17:53 - DEBUG - PIL.PngImagePlugin - STREAM b'sBIT' 54 3
Jun 23 19:17:53 jevipi inkypi[767]: 19:17:53 - DEBUG - PIL.PngImagePlugin - b'sBIT' 54 3 (unknown)
Jun 23 19:17:53 jevipi inkypi[767]: 19:17:53 - DEBUG - PIL.PngImagePlugin - STREAM b'IDAT' 69 8192
Jun 23 19:17:56 jevipi inkypi[767]: 19:17:56 - INFO - refresh_task - Updating display. | refresh_info: {'refresh_type': 'Playlist', 'playlist': 'Default', 'plugin_id': 'calendar', 'plugin_instance': 'Now', 'refresh_time': '2025-06-23T19:16:54.127849-05:00', 'image_hash': '7753a358c969fd508ecb41dd1d146d5c7fce581307e32e30738b94e73c764435'}
Jun 23 19:17:56 jevipi inkypi[767]: 19:17:56 - INFO - display.display_manager - Saving image to /home/jevi/InkyPi/src/static/images/current_image.png
Jun 23 19:17:57 jevipi inkypi[767]: 19:17:57 - DEBUG - config - Getting key orientation from config.
Jun 23 19:17:57 jevipi inkypi[767]: 19:17:57 - DEBUG - config - Getting key inverted_image from config.
Jun 23 19:17:57 jevipi inkypi[767]: 19:17:57 - DEBUG - config - Getting key resolution from config.
Jun 23 19:17:57 jevipi inkypi[767]: 19:17:57 - DEBUG - config - Getting key image_settings from config.
Jun 23 19:17:58 jevipi inkypi[767]: 19:17:58 - INFO - display.inky_display - Displaying image to Inky display.
Jun 23 19:18:09 jevipi inkypi[767]: 19:18:09 - DEBUG - config - Writing device config to /home/jevi/InkyPi/src/config/device.json
Jun 23 19:18:09 jevipi inkypi[767]: 19:18:09 - DEBUG - config - Getting key scheduler_sleep_time from config.
Jun 23 19:18:09 jevipi inkypi[767]: 19:18:09 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:09] "POST /display_plugin_instance HTTP/1.1" 200 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /playlist HTTP/1.1" 200 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /static/scripts/response_modal.js HTTP/1.1" 304 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /static/styles/main.css HTTP/1.1" 304 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /static/icons/playlist.png HTTP/1.1" 304 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /static/icons/settings.png HTTP/1.1" 304 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /images/calendar/icon.png HTTP/1.1" 304 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /static/icons/edit.png HTTP/1.1" 304 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /static/icons/display.png HTTP/1.1" 304 -
Jun 23 19:18:10 jevipi inkypi[767]: 19:18:10 - INFO - werkzeug - 192.168.1.44 - - [23/Jun/2025 19:18:10] "GET /static/icons/remove.png HTTP/1.1" 304 -
Jun 23 19:19:09 jevipi inkypi[767]: 19:19:09 - DEBUG - config - Getting key timezone from config.
Jun 23 19:19:09 jevipi inkypi[767]: 19:19:09 - INFO - refresh_task - Running interval refresh check. | current_time: 2025-06-23 19:19:09
Jun 23 19:19:09 jevipi inkypi[767]: 19:19:09 - DEBUG - config - Getting key plugin_cycle_interval_seconds from config.
Jun 23 19:19:09 jevipi inkypi[767]: 19:19:09 - INFO - refresh_task - Not time to update display. | latest_update: 2025-06-23 19:16:54 | plugin_cycle_interval: 3600
Jun 23 19:19:09 jevipi inkypi[767]: 19:19:09 - DEBUG - config - Getting key scheduler_sleep_time from config.`

fatihak commented:
Hm I'm seeing the "Displaying image to Inky display" log message so it should be updating the display
